src/vs/workbench/api/browser/positron/mainThreadLanguageRuntime.ts

```diff
@@ -958,17 +962,65 @@ class ExtHostLanguageRuntimeSessionAdapter implements ILanguageRuntimeSession {
 	}
 }
 
+enum PendingRpcStatus {
+	/**
+	 * The RPC was sent to the kernel, but the kernel didn't start working on it yet.
+	 */
+	Pending,
+	/**
+	 * The Kernel sent a state message with the RPC ID (as the parent id) and state 'Busy'
+	 * indicating that it has started handling the RPC.
+	 */
+	Executing,
+	/**
+	 * The Kernel sent a state message with the RPC ID (as the parent id) and status 'Idle'
+	 * indicating that it has finished handling the RPC.
+	 * Note: Completed means that the server finished processing the RPC, but it doesn't
+	 * mean that the RPC was successful. The promise may still be rejected.
+	 */
+	Completed,
+	/***
+	 * The Kernel closed before the RPC was completed, or the RPC response never arrived
+	 * within the timeout period.
+	 */
+	Error,
+}
+
 /**
  * Represents a pending RPC call that has been sent to the server side of a
  * comm.
  */
 class PendingRpc<T> {
 	public readonly promise: DeferredPromise<IRuntimeClientOutput<T>>;
+
+	private readonly _onDidChangeStatus = new Emitter<PendingRpcStatus>();
+	private _status: PendingRpcStatus = PendingRpcStatus.Pending;
+
+	/**
+	 * Can be used to check if the RPC is waiting to execute, or if it has already started.
+	 */
+	public readonly onDidChangeStatus: Event<PendingRpcStatus> = this._onDidChangeStatus.event;
+
+
 	constructor(
 		public readonly responseKeys: Array<string>
 	) {
 		this.promise = new DeferredPromise<IRuntimeClientOutput<T>>();
 	}
+
+	public setStatus(status: PendingRpcStatus): void {
+		this._status = status;
+		this._onDidChangeStatus.fire(status);
```

Comment:
This could use a guard to avoid firing events where the status has not changed, unless the intent is for that to be possible.

Reply:
Done in https://github.com/posit-dev/positron/pull/7286/commits/07e9d3167531f67af28c31a0e6bb119a5fae95f1